dace/autodiff/data_forwarding/store.py

```diff
@@ -0,0 +1,683 @@
+                                 backward_state: SDFGState, forward_node: nodes.AccessNode, target_node: nodes.Node,
+                                 starting_edge: dstate.MultiConnectorEdge):
+    """
+        Given the AccessNode pointing to the data required by the backward pass,
+        We will save the values of this array in a new array and forward it to the backward pass.
+        """
+
+    # Modify the forward pass to save the data in a new array
+    new_stored_array, memlets = _store_data(bwd_generator=bwd_generator,
+                                            forward_state=forward_state,
+                                            backward_state=backward_state,
+                                            forward_an=forward_node,
+                                            target_node=target_node,
+                                            edge=starting_edge)
+
+    # Check if this data needs to be forwarded through a SDFGs
+    if bwd_generator.separate_sdfgs or ad_utils.within_nested_sdfg(forward_state):
+        # We need to make sure the new array is forwarded to the backward SDFG
+        if new_stored_array.data not in bwd_generator.backward_input_arrays:
+            # If the data is needed inside a NestedSDFG
+            # This will make sure the added array is correctly forwarded
+            # and an in connector to the NestedSDFG is added
+            data_desc = new_stored_array.desc(forward_state)
+            bwd_generator.backward_input_arrays[new_stored_array.data] = data_desc
+
+    # Connect the new array to the target node
+    _connect_stored_data_to_target(bwd_generator=bwd_generator,
+                                   forward_state=forward_state,
+                                   backward_state=backward_state,
+                                   source_node=new_stored_array,
+                                   forward_node=forward_node,
+                                   starting_edge=starting_edge,
+                                   memlets=memlets,
+                                   target_node=target_node)
+
+
+def _store_data(bwd_generator: 'BackwardPassGenerator', forward_state: SDFGState, backward_state: SDFGState,
+                forward_an: nodes.AccessNode, target_node: nodes.Node,
+                edge: dgraph.MultiConnectorEdge) -> Tuple[nodes.AccessNode, List[Memlet]]:
+    """
+        Given an edge leading an AccessNode or a map to the target node in the forward state,
+        add a path from the connector for this AccessNode to store its values for all iterations.
+        This can increase the dimension of the array. i.e. the size of the stored array is
+        greater or equal to the size of the original array.
+
+        :param edge: the edge connecting the AccessNode to save data from to a map node.
+        :return: the new AccessNode which contains the stored data,
+                 a list of memlets connecting an assign tasklet to this new AccessNode.
+        """
```

Comment:
```suggestion
    Given an edge leading an AccessNode or a map to the target node in the forward state,
    add a path from the connector for this AccessNode to store its values for all iterations.
    This can increase the dimension of the array. i.e. the size of the stored array is
    greater or equal to the size of the original array.

    :param edge: the edge connecting the AccessNode to save data from to a map node.
    :return: the new AccessNode which contains the stored data,
             a list of memlets connecting an assign tasklet to this new AccessNode.
    """
```